### ISSUE TITLE
Weird behavior from `link` and `invlink` for `Distributions.Product` with DistributionsAD

### ISSUE THREAD

**Issue by torfjelde** [TuringLang/Bijectors.jl#100]
Seems to be an issue with Bijectors.jl's `compat/distributionsad.jl`:
```julia
julia> using DistributionsAD, Bijectors

julia> dists = [Gamma(2, 3) for i = 1:2]
2-element Array{Gamma{Float64},1}:
 Gamma{Float64}(α=2.0, θ=3.0)
 Gamma{Float64}(α=2.0, θ=3.0)

julia> dist = product_distribution(dists)
Product{Continuous,Gamma{Float64},Array{Gamma{Float64},1}}(v=Gamma{Float64}[Gamma{Float64}(α=2.0, θ=3.0), Gamma{Float64}(α=2.0, θ=3.0)])

julia> x = rand(dist)
2-element Array{Float64,1}:
 3.3962613472010568
 4.504150811063049

julia> Bijectors.link(dist, x)
2-element Array{Float64,1}:
 1.222675222849948
 1.5049993740830023

julia> Bijectors.invlink(dist, Bijectors.link(dist, x))
2-element Array{Float64,1}:
 1.222675222849948
 1.5049993740830023
```
while on a project with Bijectors.jl only:
```julia
julia> using Bijectors

julia> dists = [Gamma(2, 3) for i = 1:2]
2-element Array{Gamma{Float64},1}:
 Gamma{Float64}(α=2.0, θ=3.0)
 Gamma{Float64}(α=2.0, θ=3.0)

julia> dist = product_distribution(dists)
Product{Continuous,Gamma{Float64},Array{Gamma{Float64},1}}(v=Gamma{Float64}[Gamma{Float64}(α=2.0, θ=3.0), Gamma{Float64}(α=2.0, θ=3.0)])

julia> x = rand(dist)

2-element Array{Float64,1}:
 11.597241830334823
 19.19914746393061

julia> Bijectors.link(dist, x)
2-element Array{Float64,1}:
 11.597241830334823
 19.19914746393061

julia> Bijectors.invlink(dist, Bijectors.link(dist, x))
2-element Array{Float64,1}:
 11.597241830334823
 19.19914746393061

(TestingBijectors) pkg> st
Project TestingBijectors v0.1.0
Status `/tmp/TestingBijectors/Project.toml`
  [76274a88] Bijectors v0.6.6
```

Behavior is caused by the following when used with DistributionsAD.jl:
```julia
julia> @which Bijectors.link(dist, x)
link(dist::Product, x::AbstractArray{#s99,1} where #s99<:Real) in Bijectors at /home/tor/.julia/packages/Bijectors/to1z3/src/compat/distributionsad.jl:249

julia> @which Bijectors.invlink(dist, x)
invlink(d::Distribution{Multivariate,S} where S<:ValueSupport, y::Union{AbstractArray{#s25,1}, AbstractArray{#s25,2}} where #s25<:Real) in Bijectors at /home/tor/.julia/packages/Bijectors/to1z3/src/Bijectors.jl:427
```
where `src/compat/distributionsad.jl:249` will call `link` for each of the `Univariate` components of the `Product` `dist` while `src/Bijectors.jl:427` calls identity, treating `dist` as `Multivariate`.

**Comment by mohamed82008** [comment]
Btw, this issue is not fixed by the PR just merged.

**Comment by mohamed82008** [comment]
I am working on a fix.

**Comment by torfjelde** [comment]
Wait, it _didn't_ fix the fact that `invlink` had different behavior depending on whether or not `DistributionsAD` was imported?

**Comment by mohamed82008** [comment]
Ya we missed the main methods.

**Comment by torfjelde** [comment]
Fixed in #101
